src/aind_hcr_data_transformation/compress/czi_to_zarr_purezarr.py

```diff
@@ -0,0 +1,458 @@
+    """Check if dataset already exists by looking for zarr.json."""
+    fs = s3fs.S3FileSystem(anon=False)
+    return fs.exists(f"{bucket}/{key}/zarr.json")
+
+def _ensure_scale_factor_5D(scale_factor: List[int]) -> List[int]:
+    """Ensure scale_factor is a 5D list (T,C,Z,Y,X)."""
+    if len(scale_factor) > 5:
+        raise ValueError("scale_factor cannot have more than 5 dimensions")
+    return ([1] * (5 - len(scale_factor))) + scale_factor
+
+
+def _write_pyramid_level(
+    stack_group, level_idx: int, level_data, chunk_size_5d: list, 
+    base_dtype, compressor, macro_block_target_mb: int, max_workers: int
+):
+    """Write a single pyramid level in parallel-friendly way."""
+    level_shape = level_data.shape
+    level_chunks = tuple(min(c, s) for c, s in zip(chunk_size_5d, level_shape))
+    
+    # Create zarr array for this level using Zarr v3 API
+    arr_level = stack_group.create_array(
+        name=str(level_idx),
+        shape=level_shape,
+        chunks=level_chunks,
+        dtype=base_dtype,
+        compressors=[compressor],  # Use compressors parameter for v3
+        overwrite=True,
+    )
+    
+    # Write using blocked writer with optimized block size
+    block_shape = BlockedArrayWriter.get_block_shape(
+        level_data, target_size_mb=macro_block_target_mb
+    )
+    
+    logging.info(f"[purezarr] Writing level {level_idx}, shape: {level_shape}, block shape: {block_shape}")
+    BlockedArrayWriter.store(level_data, arr_level, block_shape, max_workers=max_workers)
+    
+    return arr_level
+
+
+def _write_level0_parallel(
+    czi, arr0, dataset_shape: tuple, chunk_size_5d: list, max_workers: int = 8
+) -> int:
+    """Write level 0 data using parallel chunk-aligned writes with optimized batching."""
+    z_jump = chunk_size_5d[-3]
+    total_regions = 0
+    
+    # Collect all z-blocks and their tasks upfront for better parallelization
+    all_write_tasks = []
+    
+    for z_block, axis_area in czi_block_generator(czi, axis_jumps=z_jump, slice_axis="z"):
+        z_block = pad_array_n_d(z_block)
+        z_block = np.asarray(z_block)
+        
+        z_start, z_stop = axis_area.start, axis_area.stop
+        local_z = z_stop - z_start
+        
+        # Generate all write tasks for this z-block
+        stride_y, stride_x = chunk_size_5d[-2], chunk_size_5d[-1]
+        
+        for y0 in range(0, dataset_shape[-2], stride_y):
+            y1 = min(y0 + stride_y, dataset_shape[-2])
+            for x0 in range(0, dataset_shape[-1], stride_x):
+                x1 = min(x0 + stride_x, dataset_shape[-1])
+                
+                region = (slice(0, dataset_shape[0]), slice(0, dataset_shape[1]),
+                         slice(z_start, z_stop), slice(y0, y1), slice(x0, x1))
+                sub_block = z_block[:, :, 0:local_z, y0:y1, x0:x1].copy()  # Copy to avoid reference issues
+                all_write_tasks.append((region, sub_block))
```

Comment:
I see a couple things that worry me here:

1. is the `.copy()` necessary? That's lots of allocations and memory traffic. Did you try without the copy?
2. The list might create a huge memory footprint in big/huge datasets. Can you yield the region and subblock?

Not crucial but this inner loop is O(Ny x Nx) without counting the external. We have to touch every chunk so it's fine but having these three loops hurts my heart haha. Numpy is very nice and fast, I think you can replace these python loops for numpy as_strided operations. But this is not brutally crucial, it'd help runtimes a little.